### PR TITLE
Terminate unjoined nodes

### LINKED ISSUE #107: Unjoined nodes are not rotated
In scenarios where nodes fail to join the cluster, upgrade manager will skip terminating the nodes.
This leads to a scenario where you must manually terminate nodes if you used a bad configuration when creating the node group, and you cannot upgrade your way out of it.

```
2020-07-09T17:39:07.461Z	INFO	controllers.RollingUpgrade	launch configuration name differs	{"rollingupgrade": "rollup-nodes-20200709173853"}
2020-07-09T17:39:07.461Z	INFO	controllers.RollingUpgrade	Node for instance not found	{"rollingupgrade": "rollup-nodes-20200709173853", "instanceID": "i-0b1a45f8c66d9084e"}
2020-07-09T17:39:07.461Z	INFO	controllers.RollingUpgrade	Node name for instance not found	{"rollingupgrade": "rollup-nodes-20200709173853", "instanceID": "i-0b1a45f8c66d9084e"}
2020-07-09T17:39:07.465Z	INFO	controllers.RollingUpgrade	Marked object as	{"rollingupgrade": "rollup-nodes-20200709173853", "finalStatus": "completed"}
2020-07-09T17:39:07.474Z	INFO	controllers.RollingUpgrade	Deleted the entries of ASG in the cluster store	{"rollingupgrade": "rollup-nodes-20200709173853", "asgName": "instancemgr-dev-instance-manager-my-instance-group-2"}
```

We should try to terminate the instance id even if the node is not found in the cluster and move on

### PR DESCRIPTION
**Pull request by hard-fault**
Fixes #107 

Through this fix, instances that are part of ASG but aren't part of the cluster will be allowed to terminate.

### CODE REVIEW

File: controllers/rollingupgrade_controller.go [hunk 1]
```diff
@@ -497,16 +500,14 @@ func (r *RollingUpgradeReconciler) populateNodeList(ruObj *upgrademgrv1alpha1.Ro
 
 // Loads specific environment variables for scripts to use
 // on a given rollingUpgrade and autoscaling instance
-func loadEnvironmentVariables(ruObj *upgrademgrv1alpha1.RollingUpgrade, nodeInstance *corev1.Node) error {
+func loadEnvironmentVariables(ruObj *upgrademgrv1alpha1.RollingUpgrade, instanceID string, nodeName string) error {
 	if err := os.Setenv(asgNameKey, ruObj.Spec.AsgName); err != nil {
 		return errors.New(ruObj.Name + ": Could not load " + asgNameKey + ": " + err.Error())
 	}
-	tokens := strings.Split(nodeInstance.Spec.ProviderID, "/")
```
Comment (shrinandj) [inline controllers/rollingupgrade_controller.go:504]: Doesn't this code need to move to wherever `loadEnvironmentVariables` is called from?

File: controllers/rollingupgrade_controller.go [hunk 2]
```diff
@@ -734,11 +735,13 @@ func (r *RollingUpgradeReconciler) validateNodesLaunchDefinition(ruObj *upgradem
 	ec2instances := asg.Instances
 	for _, ec2Instance := range ec2instances {
 		ec2InstanceID, ec2InstanceLaunchConfig, ec2InstanceLaunchTemplate := ec2Instance.InstanceId, ec2Instance.LaunchConfigurationName, ec2Instance.LaunchTemplate
-		if aws.StringValue(launchConfigASG) != aws.StringValue(ec2InstanceLaunchConfig) {
-			return fmt.Errorf("launch config mismatch, %s instance config - %s, does not match the asg config", aws.StringValue(ec2InstanceID), aws.StringValue(ec2InstanceLaunchConfig))
-		} else if launchTemplateASG != nil && ec2InstanceLaunchTemplate != nil {
-			if aws.StringValue(launchTemplateASG.LaunchTemplateId) != aws.StringValue(ec2InstanceLaunchTemplate.LaunchTemplateId) {
-				return fmt.Errorf("launch template mismatch, %s instance template - %s, does not match the asg template", aws.StringValue(ec2InstanceID), aws.StringValue(ec2InstanceLaunchTemplate.LaunchTemplateId))
+		if aws.StringValue(ec2Instance.LifecycleState) == "InService" {
```
Comment (shrinandj) [inline controllers/rollingupgrade_controller.go:738]: Let's change this check to get better indentation.

```
if aws.StringValue(ec2Instance.LifecycleState) != "InService" {
    continue
}
...
```

Also, is the `InService` state available as some const/enum? Direct string matching can be error prone.

File: controllers/rollingupgrade_controller.go [hunk 3]
```diff
@@ -321,6 +321,9 @@ func (r *RollingUpgradeReconciler) WaitForDesiredNodes(ruObj *upgrademgrv1alpha1
 }
 
 func (r *RollingUpgradeReconciler) WaitForTermination(ruObj *upgrademgrv1alpha1.RollingUpgrade, nodeName string, nodeInterface v1.NodeInterface) (bool, error) {
+	if nodeName == "" {
```
Comment (shrinandj) [inline controllers/rollingupgrade_controller.go:324]: Are there unit tests for this?

File: controllers/rollingupgrade_controller.go [hunk 4]
```diff
@@ -1028,14 +1031,16 @@ func (r *RollingUpgradeReconciler) DrainTerminate(
 	ch chan error) {
 
 	// Drain and wait for draining node.
-	err := r.DrainNode(ruObj, nodeName, KubeCtlCall, ruObj.Spec.Strategy.DrainTimeout)
-	if err != nil && !ruObj.Spec.IgnoreDrainFailures {
-		ch <- err
-		return
+	if nodeName != "" {
```
Comment (shrinandj) [inline controllers/rollingupgrade_controller.go:1034]: Same as the first comment here, let's add some unit tests for this.

File: controllers/rollingupgrade_controller.go (hunk 1 above)
Comment (hard-fault) [reply controllers/rollingupgrade_controller.go:504]: loadEnvironmentVariables() function is called only in the UpdateInstance() function. 
And in UpdateInstance() function the nodeName and InstanceID are already available. So, we won't need this line of code.

File: controllers/rollingupgrade_controller.go (hunk 3 above)
Comment (hard-fault) [reply controllers/rollingupgrade_controller.go:324]: Done. Added in latest commit.

File: controllers/rollingupgrade_controller.go (hunk 4 above)
Comment (hard-fault) [reply controllers/rollingupgrade_controller.go:1034]: Done. Added in latest commit.

**Review by shrinandj — APPROVED** [review]

**Comment by shrinandj** [comment]
Next time please add a `Testing Done` section to the PR.

**Pull request merged**